src/main/scala/io/iohk/ethereum/db/storage/encoding/package.scala

```diff
@@ -21,11 +21,11 @@ package object encoding {
 
   private val storedNodeEncDec = new RLPDecoder[StoredNode] with RLPEncoder[StoredNode] {
     override def decode(rlp: RLPEncodeable): StoredNode = rlp match {
-      case RLPList(nodeEncoded, references) => StoredNode(nodeEncoded, references)
+      case RLPList(nodeEncoded, references, lastUsedByBlock) => StoredNode(nodeEncoded, references, lastUsedByBlock)
```

Comment:
So DB-wise it's a backward incompatible change?

Reply:
Yes :sad:

Reply:
:/

Reply:
Let me add more info about this. In order to delay MPT node deletion we need to find a way to determine if the node can be deleted at certain point (ie, it wasn't used after that). In order to do so, the easiest way I found is to put that stamp based on block number.

Reply:
`StoredNode#lastUsedBlock`? But what's the relation of this field with reference count? How can a node be used later if its ref count drops to zero?

Reply:
Reference count has the current references the node has and last used block is the last time the block was used.

Think about the following scenario.

- `Block#0` -> `(Insert, Key1 -> V1)`
- `Block#1` -> `(Delete, Key1)`
- `Block#999` -> `(Insert, Key1 -> V1)`

Block deletion is delayed up to pruning process, so:
- Pruning `Block#0` -> Nothing done as no delete
- Pruning `Block#1` -> We have a delete but current references count is 1 as it was added at `Block#999` so block is not deleted

Now, think about this case:
- `Block#0` -> `(Insert, Key1 -> V1)`
- `Block#1` -> `(Delete, Key1)`
- `Block#998` -> `(Insert, Key1 -> V1)`
- `Block#999` -> `(Delete, Key1)`

Pruning will be:
- Pruning `Block#0` -> Nothing done as no delete
- Pruning `Block#1` -> We have a delete and current reference count is 0 because it was deleted at `Block#999` *but we cannot delete it*. Why? Because if we do so, we won't be able to provide that node to someone doing fast sync starting at `Block#998` or mining from the same one
- Pruning `Block#998` -> Nothing done as no delete
- Pruning `Block#999` -> Node has no more references and it wasn't used (999-1999), delete it from db

Reply:
Great expl, thanks